drivers: modem: Add support for quectel bg95

bwasim opened the pull request:
Adding support for Quectel BG95 Modem offloaded driver to zephyr.

The driver currently implements only the client side functions of the "socket_op_vtable", and so cannot be used for cases where Zephyr acts as a server. Moreover the driver only supports TCP for now.

Looking through the guides, the same driver should be usable for BG96 (and other modems) except for the modem boot-up sequence. Hence its named as "bg9x" instead of "bg95".

Tested extensively with Zephyr acting as MQTT endpoint and publishing / subscribing data to / from an MQTT broker.

Signed-off-by: Bilal Wasim <bilalwasim676@gmail.com>

jukkar commented:
Just wondering if these bg9x modems supports PPP mode instead of offloaded one, any idea?

bwasim commented:
> Just wondering if these bg9x modems supports PPP mode instead of offloaded one, any idea?

They do - I actually did try PPP on this and was able to get the modem booted, but PPP failed at the network layer protocol stage, and eventually the link got terminated. Didn't look too deep into it, but I was using CAT-M1 SIM and it "seemed" that Zephyr PPP only supported GSM.

jukkar commented:
> Didn't look too deep into it, but I was using CAT-M1 SIM and it "seemed" that Zephyr PPP only supported GSM.

Yes, it would be great if we could improve the ppp support in the gsm driver. I have not used CAT-M1 SIM so do not know what kind of changes would be needed there.
Anyway, thanks for the info.

mniestroj commented:
I used PPP with BG96 and it worked really well. The thing is that I didn't use gsm_ppp layer so far, but some custom AT command layer instead. I have also access to project with BG77, which is also a Cat M1, and there was no issues either. It is only a matter of fine tuning AT commands.

jukkar commented:
> It is only a matter of fine tuning AT commands.

This is what I suspected, it seems that each modem manufacturer wants to have slightly different way of doing things.

galak commented:
Can you add this to `tests/drivers/build_all`. 

You can see following commit for patterrn:


commit dc0466f8a168553954052c299b383cf95555b849
Author: Kumar Gala <kumar.gala@linaro.org>
Date:   Thu Oct 15 03:23:06 2020 -0500

    tests: drivers: build_all: Rework modem test to be more generic

galak commented on drivers/modem/Kconfig.quectel-bg9x:
```diff
@@ -0,0 +1,68 @@
+	  Choose this setting to enable quectel BG9x LTE-CatM1/NB-IoT modem
+	  driver.
+
+config MODEM_QUECTEL_BG9X_NAME
```
Is there a reason this needs to be a Kconfig?

bwasim replied:
Yeah, doesn't seem to be necessary - Was a copy-paste from the SARA-R4 driver. Will remove..

galak commented on drivers/modem/Kconfig.quectel-bg9x:
```diff
@@ -0,0 +1,68 @@
+	  context.  This value is specific to the network provider and may
+	  need to be changed.
+
+config MODEM_QUECTEL_BG9X_PASSWORD
```
hmm, seems like bad security if you have to hardcode the password ;)

bwasim replied:
This is the APN password which is generally made publicly available by the SIM vendor anyways. So I don't think this is a problem..

See https://soracom.zendesk.com/hc/en-us/articles/218427317-What-is-an-APN-setting-and-how-do-you-set-it-up- for example.

parthitce commented on drivers/modem/Kconfig.quectel-bg9x:
```diff
@@ -0,0 +1,68 @@
+	select MODEM_CMD_HANDLER
+	select MODEM_IFACE_UART
+	select MODEM_SOCKET
+	select NET_OFFLOAD
```
Is it offloading net_context?

bwasim replied:
Nope, its offloading Net Sockets (NET_SOCKETS_OFFLOAD)..

parthitce replied:
`NET_OFFLOAD` can this be removed?

parthitce commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1118 @@
+
+		/* process, then yield. */
+		mctx.cmd_handler.process(&mctx.cmd_handler, &mctx.iface);
+		k_yield();
```
k_sem_take will re-sched the thread if I am not wrong. Does this need to yield explicit?

bwasim replied:
Yup you're right. This was a copy-paste from the Sara-R4 driver, and I'm not sure why that driver resorted to an explicit yield.

parthitce commented on drivers/modem/Kconfig.quectel-bg9x:
```diff
@@ -0,0 +1,68 @@
+	  to the rest of the network stack, letting the rx thread continue
+	  processing data.
+
+config MODEM_QUECTEL_BG9X_APN
```
Is there a way to configure APN during runtime?

bwasim replied:
Not yet, Will consider adding it in the next revision.

parthitce reviewed (commented):
Nitpicks: conditional and loop constructs needs to be enclosed with `{}`

parthitce commented:
Also I couldn't compile the driver,

```
zephyr/drivers/modem/quectel-bg9x.c: In function 'modem_init':
zephyr/drivers/modem/quectel-bg9x.c:1067:18: error: 'struct modem_iface_uart_data' has no member named 'isr_buf'
 1067 |  mdata.iface_data.isr_buf       = &mdata.iface_isr_buf[0];
      |                  ^
zephyr/drivers/modem/quectel-bg9x.c:1068:19: error: 'struct modem_iface_uart_data' has no member named 'isr_buf_len'; did you mean 'rx_rb_buf_len'?
 1068 |  mdata.iface_data.isr_buf_len   = sizeof(mdata.iface_isr_buf);
      |                   ^~~~~~~~~~~
      |                   rx_rb_buf_len
[107/163] Building C object zephyr/subsys/net/ip/CMakeFiles/subsys__net__ip.dir/net_context.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build zephyr/build
```

Please check 763bd27c7599c78ed9b77867dab29553e0781ddf

bwasim commented:
> Also I couldn't compile the driver,
> 
> ```
> zephyr/drivers/modem/quectel-bg9x.c: In function 'modem_init':
> zephyr/drivers/modem/quectel-bg9x.c:1067:18: error: 'struct modem_iface_uart_data' has no member named 'isr_buf'
>  1067 |  mdata.iface_data.isr_buf       = &mdata.iface_isr_buf[0];
>       |                  ^
> zephyr/drivers/modem/quectel-bg9x.c:1068:19: error: 'struct modem_iface_uart_data' has no member named 'isr_buf_len'; did you mean 'rx_rb_buf_len'?
>  1068 |  mdata.iface_data.isr_buf_len   = sizeof(mdata.iface_isr_buf);
>       |                   ^~~~~~~~~~~
>       |                   rx_rb_buf_len
> [107/163] Building C object zephyr/subsys/net/ip/CMakeFiles/subsys__net__ip.dir/net_context.c.obj
> ninja: build stopped: subcommand failed.
> FATAL ERROR: command exited with status 1: /usr/bin/cmake --build zephyr/build
> ```
> 
> Please check [763bd27](https://github.com/zephyrproject-rtos/zephyr/commit/763bd27c7599c78ed9b77867dab29553e0781ddf)

Thanks for pointing this out, Will fix. To date, this driver has been tested only with Zephyr v2.4.0 and hence this got missed.

bwasim commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1118 @@
+				     NULL, 0U, "AT+QIACT=1", &mdata.sem_response,
+						 MDM_CMD_TIMEOUT);
+	if (ret < 0)
+		goto restart;
```
Added an upper bound on the number of retries..

WilliamGFish commented:
> I used PPP with BG96 and it worked really well. The thing is that I didn't use gsm_ppp layer so far, but some custom AT command layer instead. I have also access to project with BG77, which is also a Cat M1, and there was no issues either. It is only a matter of fine tuning AT commands.

Can you share how you did this and the "fine tuned AT Commands" that you used to make the PPP function correctly. It would be much appricated.

Billy..

bwasim commented:
@mike-scott / @jukkar / @galak , Can you kindly review this pull request ? Thanks..

jukkar commented on drivers/modem/Kconfig.quectel-bg9x:
```diff
@@ -0,0 +1,64 @@
+
+config MODEM_QUECTEL_BG9X_APN
+	string "APN for establishing network connection"
+	default "hologram"
```
This value seems to be specific to your network. Perhaps `internet` would be more generic default value here

jukkar commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1147 @@
+ */
+static void socket_close(struct modem_socket *sock)
+{
+	char buf[32];
```
Instead of hard coded value, we could use a template for expected input here. So something like
`char buf[sizeof("AT+QICLOSE=xx")];`
assuming here that the sock_fd is <100.

jukkar commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1147 @@
+				k_timeout_t timeout)
+{
+	int  ret;
+	char send_buf[32] = {0};
```
here too, I would prefer something like `char send_buf[sizeof("AT+QISEND=xx,xxx")` here

jukkar commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1147 @@
+				socklen_t *fromlen)
+{
+	struct modem_socket *sock = (struct modem_socket *)obj;
+	char   sendbuf[32]	  = {0};
```
ditto

jukkar commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1147 @@
+{
+	struct modem_socket *sock     = (struct modem_socket *) obj;
+	uint16_t	    dst_port  = 0;
+	char		    *protocol = "TCP", buf[64] = {0};
```
please use use a template for buf size

jukkar commented on drivers/modem/quectel-bg9x.h:
```diff
@@ -0,0 +1,190 @@
+	return h;
+}
+
+static int net_offload_dummy_get(sa_family_t family, enum net_sock_type type,
```
This function should probably be placed in .c file?

jukkar reviewed (changes requested):
Minor nits but looks quite resonable, the core functionality seem be coming from ublox-sara modem driver anyway.

jukkar commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1155 @@
+	.setsockopt	= NULL,
+};
+
+static struct net_offload modem_net_offload = {
+	.get = net_offload_dummy_get,
+};
```
If we disable NET_OFFLOAD, then this should not be needed

bwasim replied:
Yup, will remove. I based this driver on v2.4.0 where removing NET_OFFLOAD caused build failure. This failure was fixed under https://github.com/zephyrproject-rtos/zephyr/pull/28807

jukkar commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1155 @@
+	const struct device *dev = net_if_get_device(iface);
+	struct modem_data *data	 = dev->data;
+
+	/* Direct socket offload used instead of net offload: */
+	iface->if_dev->offload = &modem_net_offload;
```
and we should not need these if NET_OFFLOAD is not set.

Just wondering if there is some side effect which requires that NET_OFFLOAD is set. If this is the case, then it sounds like a bug (because we do not use NET_OFFLOAD for anything). Can you try to remove the NET_OFFLOAD and see what happens?

jukkar commented on drivers/modem/quectel-bg9x.h:
```diff
@@ -0,0 +1,180 @@
+		buf[i] = data[i];
+	}
+
+	return atoi(buf);
```
Could we use the `modem_atoi()` here as it checks the value of the string properly?

parthitce commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1155 @@
+
+	default:
+		errno = EINVAL;
+		return -1;
```
`return -EINVAL` will work?

bwasim replied:
It will work, but I think the standard posix behavior is to return errors in `errno` variable, though this function doesn't seem to be POSIX compliant.

parthitce reviewed (commented):
Nitpick: `errno` needs to be negative if I am not wrong.

parthitce commented on drivers/modem/quectel-bg9x.h:
```diff
@@ -0,0 +1,180 @@
+#define MDM_SENDMSG_SLEEP		  K_MSEC(1)
+#define MDM_MAX_DATA_LENGTH		  1024
+#define MDM_RECV_MAX_BUF		  30
+#define MDM_RECV_BUF_SIZE		  1024
```
`NET_BUF_POOL_DEFINE` allocates 30 * 1024 for the `mdm_recv_pool`. Can this be reduced?

bwasim replied:
May be the best solution is to put either of these values in Kconfig ? This will allow users to make modifications based on their needs without modifying the source code, And there's no reason why reducing the size of pool won't work..

bwasim commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1155 @@
+		}
+
+		counter++;
+		k_sleep(K_SECONDS(2));
```
I think the problem is not that it will wait 100 seconds for the modem to respond to `AT` commands, but that it will basically hang the system for this time. May be moving this code from the main thread to its own worker thread is a better solution ? Once the modem is initialized (in the background), it can send an event to application so that the app can start using this network device. I think PPP already works this way, but not sure. 

In any case, I think I will want to do a separate PR to fix this problem, if we agree to a solution.

parthitce commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1141 @@
+#include "quectel-bg9x.h"
+
+static struct k_thread	       modem_rx_thread;
+static struct k_work_q	       modem_workq;
```
@bwasim Also I am not sure about the purpose of this workqueue. I see it's created and never consumed or no work/delayed work is posted to this queue. Is it consumed from modem layer or somewhere else?

bwasim replied:
This is used to get the RSSI continuously during the Modem Operation. See modem_setup / modem_rssi_query_work for details.

bwasim committed:
drivers: modem: Add support for quectel bg95

Adding support for Quectel BG95 Modem offloaded driver
to zephyr.

The driver currently implements only the
client side functions of the "socket_op_vtable", and
so cannot be used for cases where Zephyr acts as a
server. Moreover the driver only supports TCP for now.

Looking through the guides, the same driver should be
usable for BG96 (and other modems) except for the modem
boot-up sequence. Hence its named as "bg9x" instead of
"bg95".

Tested extensively with Zephyr acting as MQTT endpoint
and publishing / subscribing data to / from an MQTT
broker.

Signed-off-by: Bilal Wasim <bilalwasim676@gmail.com>

bwasim commented:
Seeing strange CI failures which have nothing to do with the PR itself.. 
```
HEAD is now at c39888f Update unsupported commands test to handle command status response
--
  | Previous HEAD position was 83c964e CMakeLists.txt: Rearrange library creation to be more flexible
  | HEAD is now at 6edaaaa CMakeLists.txt: Build TF-M based on new Kconfig options
  | ERROR: update failed for project ci-tools
  | touch: cannot touch 'sanity-out/sanitycheck.xml': No such file or directory
  | 🚨 Error: The command exited with status 1
```
Closing / Pre-opening to re-trigger CI.

bwasim commented:
ping !

@jukkar @galak @mike-scott @parthitce

jukkar reviewed (commented):
Looks good. Can you check whether the #30149 would require changes here as this PR is somewhat related to ublox-sara modem driver.

bwasim commented:
> Looks good. Can you check whether the #30149 would require changes here as this PR is somewhat related to ublox-sara modem driver.

Thanks, I will constify the required stuff in the driver.

XavierChapron commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -0,0 +1,1156 @@
+}
+
+/* Register the device with the Networking stack. */
+NET_DEVICE_OFFLOAD_INIT(modem_gb9x, "MODEM_BG9X",
```
I think we shoud rather replace `"MODEM_BG9X"` by  `DT_INST_LABEL(0)`

XavierChapron commented on drivers/modem/quectel-bg9x.h:
```diff
@@ -0,0 +1,166 @@
+#endif
+};
+
+static inline int digits(int n)
+{
+	int count = 0;
+
+	while (n != 0) {
+		n /= 10;
+		++count;
+	}
+
+	return count;
+}
+
+static inline uint32_t hash32(char *str, int len)
+{
+#define HASH_MULTIPLIER		37
+
+	uint32_t h = 0;
+	int i;
+
+	for (i = 0; i < len; ++i) {
+		h = (h * HASH_MULTIPLIER) + str[i];
+	}
+
+	return h;
+}
```
I don't think theses should be in the header file.
I would have put them in the c file.

bwasim replied:
No real preference - Moved !

mike-scott reviewed (approved):
LGTM
